src/wrappers/python/PyOpenEXR.cpp

```diff
@@ -0,0 +1,2453 @@
+#include <ImfRationalAttribute.h>
+#include <ImfStringAttribute.h>
+#include <ImfStringVectorAttribute.h>
+#include <ImfFloatVectorAttribute.h>
+#include <ImfTileDescriptionAttribute.h>
+#include <ImfTimeCodeAttribute.h>
+#include <ImfVecAttribute.h>
+
+#include <typeinfo>
+#include <sys/types.h>
+
+namespace py = pybind11;
+using namespace py::literals;
+
+using namespace OPENEXR_IMF_NAMESPACE;
+using namespace IMATH_NAMESPACE;
+
+extern bool init_OpenEXR_old(PyObject* module);
+
+namespace pybind11 {
+namespace detail {
+
+    // From https://github.com/AcademySoftwareFoundation/OpenImageIO/blob/master/src/python/py_oiio.h
+    //
+    // This half casting support for numpy was all derived from discussions
+    // here: https://github.com/pybind/pybind11/issues/1776
+
+    // Similar to enums in `pybind11/numpy.h`. Determined by doing:
+    // python3 -c 'import numpy as np; print(np.dtype(np.float16).num)'
+    constexpr int NPY_FLOAT16 = 23;
+
+    template<> struct npy_format_descriptor<half> {
+        static pybind11::dtype dtype()
+        {
+            handle ptr = npy_api::get().PyArray_DescrFromType_(NPY_FLOAT16);
+            return reinterpret_borrow<pybind11::dtype>(ptr);
+        }
+        static std::string format()
+        {
+            // following: https://docs.python.org/3/library/struct.html#format-characters
+            return "e";
+        }
+        static constexpr auto name = _("float16");
+    };
+
+}  // namespace detail
+}  // namespace pybind11
+
+namespace {
+
+#include "PyOpenEXR.h"
+
+//
+// Create a PyFile out of a list of parts (i.e. a multi-part file)
+//
+
+PyFile::PyFile(const py::list& parts)
+    : parts(parts)
+{
+    int part_index = 0;
+    for (auto p : this->parts)
+    {
+        if (!py::isinstance<PyPart>(p))
+            throw std::invalid_argument("must be a list of OpenEXR.Part() objects");
+
+        PyPart& P = p.cast<PyPart&>();
+        P.part_index = part_index++;
+    }
+}
+
+//
+// Create a PyFile out of a single part: header, channels,
+// type, and compression (i.e. a single-part file)
+//
+
+PyFile::PyFile(const py::dict& header, const py::dict& channels)
+{
+    parts.append(py::cast<PyPart>(PyPart(header, channels, "")));
+}
+
+//
+// Read a PyFile from the given filename.
+//
+// Create a 'Part' for each part in the file, even single-part files. The API
+// has convenience methods for accessing the first part's header and
+// channels, which for single-part files appears as the file's data.
+//
+// By default, read each channel into a numpy array of the appropriate pixel
+// type: uint32, half, or float.
+//
+// If 'rgba' is true, gather 'R', 'G', 'B', and 'A' channels and interleave
+// them into  a 3- or 4- (if 'A' is present) element numpy array. In the case
+// of raw 'R', 'G', 'B', and 'A' channels, the corresponding key in the
+// channels dict is "RGB" or "RGBA".  For channels with a prefix,
+// e.g. "left.R", "left.G", etc, the channel key is the prefix.
```

Comment:
I would like to have another thought about this API. 

What happens with the `rgba` parameter is essentially this:

1. There is an extra flip in the core Python API C implementation.
2. If you do NOT turn it on, what you get is more low level (more conforming to the file's storage format), but more difficult to use in subsequent python/numpy operations.
3. If you DO turn it on, what you get is higher level, but easier to use in subsequent numpy operations, especially for beginners.
4. This data transformation can be done in python with numpy, without bothering with C++ code. 

It caught my attention because:
1. The default is hard to use and more low level, but is general purpose. You have to flip a switch to get something that's easier to use for the noob use cases, but will fail in advanced cases. This feels backwards.
2. This can be done with numpy calls _after_ the binding layer, instead of doing it in C++ _before_ the binding layer 

I suggest we change it to one of the following:
1. Remove this parameter and the channel gathering in C++ code. Provide another OpenEXR.utils module that is entirely written in python that does tasks similar to this. (that can be either inside or outside of the official repo -- we can debate whether that would reinvent wheels of openimageIO)   
2. Flip the semantics this parameter, the C++ code will always attempt to gather channel into a nice `rgba` layout if it could do so, but you can put in an extra parameter to tell it to NOT do this.

Reply:
I rethought this after a week.

I think the most striaght forward way might be simply

1. don't have this parameter, simplify the API interface
2. always prep a channel "RGB" and "RGBA" numpy array if possible according to the file structure. If not possible, then don't prep it.

There are some python techniques that can do a lazy bind (e.g. only allocate memory on R, G, B initially and only allocate "RGBA" when this key is queried on the channel dict), but i wonder if that worth it. 

Overall i think my thought is about that extra parameter. either remove it or move it onto the channel object. It probably should not stay on the `File` object's initializer

Reply:
I tend to agree.  I've left it in place but inverted the sense of the parameter and renamed it `separate_channels`, off by default. So the default behavior is the most convenient, even if it differs from the file structure.

I'm still not completely happy with this.